ST/40_exception_handling_and_custom_exceptions/FileOperationExercise.java

```diff
@@ -0,0 +1,45 @@
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.logging.Logger;
+
+public class FileOperationExercise {
+
+    private static final Logger logger = Logger.getLogger(FileOperationExercise.class.getName());
+
+    //Run this method by clicking on Run FileOperationExercise.main()
+    public static void main(String[] args) {
+        try {
+            //Change fileName to get IOException
+            performFileOperations("ST/40_exception_handling_and_custom_exceptions/sample_file.txt");
+        } catch (IOException e) {
+            logger.severe("IOException occurred: " + e.getMessage());
+        } catch (InvalidDataException e) {
+            logger.severe("InvalidDataException occurred: " + e.getMessage());
+        } catch (ArithmeticException e) {
+            logger.severe("ArithmeticException occurred: " + e.getMessage());
+        }
+    }
+
+    private static void performFileOperations(String fileName) throws IOException, InvalidDataException {
+        try (BufferedReader reader = new BufferedReader(new FileReader(fileName))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                int number = convertToInteger(line);
+                // Performs division on 0 if provided value "10" - throws an ArithmeticException
+                if (number == 10) {
+                    int divideOnZero = number / 0;
+                }
+            }
+        }
+    }
+
+    //This method is trying to parse value from file to Integer , if fails throws your InvalidDataException
+    private static int convertToInteger(String input) throws NumberFormatException, InvalidDataException {
+        try {
+            return Integer.parseInt(input);
+        } catch (NumberFormatException e) {
+            throw new InvalidDataException("Invalid data format: " + input);
```

Comment:
can we have separate example of using `throws `and separate example of using `throw new`?
Just to show that by using throws, the exception is passed to the next method for handling and when we using throw, the exception is handled right in the implemented method

Reply:
we have it in previews method ,  in case of 10 value -> ArithmeticExeption
![image](https://github.com/bodacd/automation-study-materials/assets/156786783/89c2e350-2473-4480-a68c-a674bf5fe4f2)